### PR TITLE
Answer tests on Travis

### PR DESCRIPTION
**Pull request by git-abhishek**
## PR Background

Currently all the answer testing is done at Jenkins. Many times Jenkins service is down and takes a few days to check the issue and bring it up. This hampers developers' time as they can not run answer test suite against their changes and thus, lot of development gets blocked.

With this PR, I plan to reduce the dependency on Jenkins by enabling answer testing on cloud platforms like Travis. These platforms have high service up-time.

Answer testing on Travis is possible if we test using fake datasets instead of real datasets (otherwise huge network I/O). Since, many modules (except `frontends`) of yt can be tested with only fake datasets, this makes it possible to have a subset of answer tests that could be run on cloud services.

Thus, it makes sense to have answer tests running on high availability cloud platforms and reduce total reliance on Jenkins.

## PR Summary
Using yt answer testing, an attempt to run answer tests on Travis is being done. Following are the highlights of this PR:

- Remove real data for testing in `yt/visualization/tests/test_line_plots.py` and add answer testing with fake dateset
- Division and Running of tests:
  - Divide the tests into unit test and answer tests using nose's `@attr("answer_test")` tag
  - Run tests excluding the tests with the `answer_testing` tag
  - ~Read `tests/cloud_answer_tests.yaml` to get the list of all the answer test and execute these tests~
  -~[ ] **TODO**: A script could be written that would read yt source files and based on `@attr("answer_test")` tag would generate the file `tests/cloud_answer_tests.yaml`. This would remove the need to manually update the `tests/cloud_answer_tests.yaml` file for each new answer test.~

- Answer Testing
  - Ability to run answer test on Travis (currently doing it for py27 and py36 version as done by yt's Jenkins)
  - If no gold answer is present in answer-store, upload new answers to the upload service ~(currently using [cloudinary](https://cloudinary.com/))~ **yt's curldrop server** and finally fail the tests
  - If the answer tests fail, generate and upload a html page listing all the failed tests with their actual, expected and difference plot (eg for a dummy run [index.html](http://use.yt/upload/560ed34f) got generated).

- #### ~Open Tasks~
  ~Currently I am using [cloudinary](https://cloudinary.com/) for hosting [html](https://res.cloudinary.com/dz58gsc4u/raw/upload/v1530975515/yt/2018-07-09/index_auc6pt.html) and dumping [new answer test files](https://res.cloudinary.com/dz58gsc4u/raw/upload/v1531115627/yt/2018-07-09/answers/new-answers_vxspme.zip). However, I am open to discussion for changing to a different service like curldrop or any other solution.~ 
~**Note:** For this PR I created a new (temporary and personal) cloudinary account. Currently I have updated Travis repository variables to store the key of this account. If we are going forward with cloudinary as a service, then we need to create a new yt account of cloudinary and update the Travis key-value pairs appropriately. If we discard this service, I will delete the keys that I created.~

Note: The first run will fail and upload the answers for the new answer tests added. Once the new answers are added to the answer store subsequent runs should pass.

## PR Checklist

- [X] Code passes flake8 checker
- [X] New features are documented, with docstrings and narrative docs

## Adding Reviewers
@ngoldbaum @Xarthisius @colinmarc

### CODE REVIEW

File: tests/nose_runner.py [hunk 1]
```diff
@@ -93,7 +109,339 @@ def generate_tasks_input():
...
+        expected and difference plot file locations of the test.
+
+    Returns
+    -------
+    string
+        a html page
+
+    """
+
+    html_template = """<html><head>
```
Comment (colinmarc) [inline tests/nose_runner.py]: I'd put a newline after the quotes.

File: tests/nose_runner.py [hunk 2]
```diff
@@ -93,7 +109,339 @@ def generate_tasks_input():
...
+                encoded_images[key] = img_data
+
+        formatted_row = row_template.format(encoded_images["Actual"],
+                                            encoded_images["Expected"],
+                                            encoded_images["Difference"],
+                                            test_name)
+        rows.append(formatted_row)
+
+    body = table_template.format(rows='\n'.join(rows))
+    html = html_template.format(header="Failed Answer Tests", body=body)
```
Comment (colinmarc) [inline tests/nose_runner.py]: You can combine `table_template` and `body_template` to make this one statement. (And there's no need to parameterize `header`).

Comment (colinmarc) [reply tests/nose_runner.py]: Also, maybe I'm lacking context, but I feel like 'Failed Answer Tests' isn't super descriptive. A helpful paragraph before the results might be appropriate, too.

File: tests/nose_runner.py [hunk 3]
```diff
@@ -93,7 +109,339 @@ def generate_tasks_input():
...
+    return html
+
+def upload_to_curldrop(data, filename):
+    """Uploads file to yt's curldrop server
```
Comment (colinmarc) [inline tests/nose_runner.py]: Does this function exist nowhere else in the whole library? What's using curldrop now?

File: tests/nose_runner.py [hunk 4]
```diff
@@ -93,7 +109,339 @@ def generate_tasks_input():
...
+        argv = []
+        argv += base_argv
+        argv.append('--answer-name=%s' % "{}_{}".format(pyver, answer))
+        argv += tests["answer_tests"][answer]
+        args.append(argv)
+
+    return args
+
+
+def generate_webpage(failed_answers):
```
Comment (colinmarc) [inline tests/nose_runner.py]: This needs a more specific name.

File: tests/nose_runner.py [hunk 5]
```diff
@@ -93,7 +109,339 @@ def generate_tasks_input():
...
+        Response as returned by the upload service
+
+    """
+    html = generate_webpage(failed_answers)
+    # convert html str to bytes
+    html = html.encode()
+    timestamp = datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+    filename = "failed_answers_{}.html".format(timestamp)
+
+    response = upload_to_curldrop(data=html, filename=filename)
```
Comment (colinmarc) [inline tests/nose_runner.py]: I'd inline `upload_to_curldrop` here, if it's really only used once.

File: tests/nose_runner.py [hunk 6]
```diff
@@ -93,7 +109,339 @@ def generate_tasks_input():
...
+    return response
+
+def generate_missing_answers(answer_dir, missing_answers):
+    """Generate golden-answers
```
Comment (colinmarc) [inline tests/nose_runner.py]: Super tiny nitpick, but hyphenating doesn't make sense here or in the rest of the description.

File: tests/nose_runner.py [hunk 7]
```diff
@@ -93,7 +109,339 @@ def generate_tasks_input():
...
+        directory location to save the generated answers
+
+    missing_answers : list of list of string
+        Collection of missing answer tests, where the inner list of string
+        specifies the user given name of the answer test and the corresponding
+        test function.
+        eg. [['--answer-name=answers_multi_line_plot',
+         'yt/visualization/tests/test_line_plots.py:test_multi_line_plot']]
+
+    Returns
```
Comment (colinmarc) [inline tests/nose_runner.py]: This should probably just raise if it fails for some reason.

File: yt/visualization/tests/test_line_plots.py [hunk 8]
```diff
@@ -34,48 +32,62 @@ def image_from_plot(filename_prefix):
     test.prefix = test_prefix
     return test
 
-tri2 = "SecondOrderTris/RZ_p_no_parts_do_nothing_bcs_cone_out.e"
-
-@requires_ds(tri2)
```
Comment (colinmarc) [inline yt/visualization/tests/test_line_plots.py]: Can giving this a generated dataset be done in a separate PR? It seems orthogonal.

**Comment by git-abhishek** [comment]
Replying to @colinmarc 
>Before I get too deep into the details of the code, I have a few questions about approach:
>
>- Why the division between answer tests and other tests? 

Division is mainly due to different commands required to run the tests. While all the unittests could just be run simply using `coverage run $(which nosetests) -c nose.cfg`, answer tests require specific parameters in the command to execute like `coverage run $(which nosetests) --local-dir="answer-store" --with-answer-testing --local --answer-name=answers_line_plot yt/visualization/tests/test_line_plots.py:test_line_plot`. The last part of this command `--answer-name=answers_line_plot yt/visualization/tests/test_line_plots.py:test_line_plot` will vary for each test.

>- And why do they have to be run under subprocess? What's wrong with the way nose runs them normally? If the issue is just figuring out which ones failed, is there another way to get that information out of nose? For example, by defining a plugin and overriding [this method](http://nose.readthedocs.io/en/latest/plugins/interface.html#nose.plugins.base.IPluginInterface.handleFailure).

As mentioned by Ned Batchelder, in the [FAQ](https://coverage.readthedocs.io/en/coverage-4.5.1a/faq.html#frequently-asked-questions) of Coverage, "the best way to use nose and coverage.py together is to run nose under coverage.py : `coverage run $(which nosetests)`". Further, I observed discrepancies while running coverage with nose api. Hence to make coverage and nose work correctly, I have chosen subprocess.


> - Why do golden answers have to be uploaded? Can't they just be committed to the repo?

Golden answers are committed to the repo. 
When the golden answer for a particular test is not found (which will happen when a new answer test is written and its golden answer is not yet pushed) only then it is uploaded using curldrop. This is done so that the developer need not worry about generating answers for different environments (like py27 and py36) and could just focus on writing new answer tests. Once it is uploaded, the developer can download it from the curldrop url, validate answers manually and commit to the repo. The subsequent runs of the answer test will not upload answers to curldorp.

 >- Do you want to let people running the tests locally upload to curldrop? What about letting them see >the HTML file locally instead?

I think, this is a good idea to allow viewing HTML file locally without being uploaded to curldrop. Should I check for environment variables like `TRAVIS` before uploading and if running locally open the HTML in web browser?
Same thing could also be done for generating new golden answers?

*Let me know if I failed to explain any point or it needs more explanation.*

**Comment by colinmarc** [comment]
> Division is mainly due to different commands required to run the tests.

As the only unique argument is the name of the test, I don't think that's accurate? It seems like what you're doing is reading the output to figure out whether the test failed, so you can generate the HTML if it did. If you can hook into nose a different way, and just let nose run the tests for you alongside all the other ones, you can get rid of the yaml stuff and the command line/subprocess parts, which are the fiddliest parts of the PR.

> This is done so that the developer need not worry about generating answers for different environments (like py27 and py36) and could just focus on writing new answer tests.

Is that something we already do on Jenkins? If not, that's exactly the kind of quality-of-life improvement I would put in a second PR.

> Should I check for environment variables like TRAVIS before uploading and if running locally open the HTML in web browser?

That's a good way to do it, yeah. Printing a link to the output but having a `file:///` link is probably friendlier than opening a web browser, but either way.

However, that was just one part to my question. There's also a 'security' implication; can users upload to curldrop if they set an environment variable or pass a command line option? It's worth thinking about, even if the risk isn't that great.

**Comment by Xarthisius** [comment]
@colinmarc I can answer some of the question you raised here:

> Why the division between answer tests and other tests?
Purely because of the difference in required resources. We even had `--answer-big` to distinguish between something that can be run on a decent laptop and other more resource hungry tests.

> And why do they have to be run under subprocess? What's wrong with the way nose runs them normally? 
It was required to parallelize tests to bring them down to decent walltime. Nose `-P` simply doesn't work properly in this case: tests were failing (most likely poorly written fixtures on our end), total walltime was even longer than running them on single process.

**Comment by ngoldbaum** [comment]
For the example on the last image, can you make the link more visually distinctive? I had to stare at that screenshot for about 30 seconds before I spotted it at the bottom. Maybe along with a message that says "more details about the test failure can be found at the following URL:". Other ways to make something more visually stand out is to surround it with brackets and creative applications of emoji :)

File: yt/visualization/tests/test_line_plots.py (hunk 8 above)
Comment (git-abhishek) [reply yt/visualization/tests/test_line_plots.py]: These tests modifications are included in this PR, just to test whether this PR works or not? If I change these tests in another PR, then testing this PR would be quite difficult.

File: yt/visualization/tests/test_line_plots.py [hunk 9]
```diff
@@ -34,48 +32,62 @@ def image_from_plot(filename_prefix):
...
+                       field_labels=field_labels)
+    plot.annotate_legend(fields[0])
+    plot.annotate_legend(fields[1])
+    plot.set_x_unit('cm')
+    plot.set_unit(fields[0], 'kg/cm**3')
+    plot.annotate_title(fields[0], "Density Plot")
     yield compare(ds, plot, "answers_line_plot")
 
-@requires_ds(tri2)
+@attr("answer_test")
```
[the remark this answers is not in the document]
Comment (git-abhishek) [reply yt/visualization/tests/test_line_plots.py]: Yes, that's correct. However, we also need to update each individual tests with a `fake_dataset`.

File: tests/tests.yaml [hunk 10]
```diff
@@ -74,8 +74,6 @@ answer_tests:
     - yt/visualization/tests/test_mesh_slices.py:test_tri2
     - yt/visualization/tests/test_mesh_slices.py:test_quad2
     - yt/visualization/tests/test_mesh_slices.py:test_multi_region
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply tests/tests.yaml]: OK cool. Let's try really hard to not leave things in a state where we've partially converted things to run on travis when you're done with your internship.

File: yt/fields/tests/test_fields_plugins.py [hunk 11]
```diff
@@ -25,38 +25,37 @@ def myfunc():
 foobar = 12
 '''
 
+class TestPluginFile(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply yt/fields/tests/test_fields_plugins.py]: I've merged the other PR :)

**Comment by ngoldbaum** [comment]
The rest of the user experience looks great, very clear :)

File: tests/cloud_answer_tests_runner.py [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+    -------
+    requests.models.Response
+        Response as returned by the upload service
+
+    """
+    html = generate_failed_answers_html(failed_answers)
+    # convert html str to bytes
+    html = html.encode()
+    timestamp = datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+    filename = "failed_answers_{}.html".format(timestamp)
```
Comment (colinmarc) [inline tests/cloud_answer_tests_runner.py]: Can you grab the run ID from travis and use that in the filename instead?

File: tests/cloud_answer_tests_runner.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+
+    if args.upload_failed_tests or args.upload_missing_answers:
+        failed_answers, missing_answers = parse_nose_xml(NOSETEST_XML)
+
+    if args.upload_failed_tests and failed_answers:
+        response = upload_failed_answers(failed_answers)
+        if response.ok:
+            msg = " Successfully uploaded failed answer tests result. Url: "
+            url = response.text.split("\n")[1]
+            log.info(COLOR_BLUE + msg + url + COLOR_RESET)
```
Comment (colinmarc) [inline tests/cloud_answer_tests_runner.py]: 🏳️‍🌈

File: tests/cloud_answer_tests_runner.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+        img_path[key] = result.group(1)
+
+    if not unknown_failure:
+        return img_path
+    return None
+
+def parse_nose_xml(nose_xml):
+    """Parse xml file generated by nosetests.
+
+    Parse nose xml file to find following details:
```
Comment (colinmarc) [inline tests/cloud_answer_tests_runner.py]: Weird indent

File: tests/cloud_answer_tests_runner.py [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+        for error in testcase:
+            test_name = testcase.attrib["classname"] + ":" \
+                        + testcase.attrib["name"]
+            if "No such file or directory" in error.attrib["message"]:
+                missing_answers.append(test_name)
+            elif "Items are not equal" in error.attrib["message"]:
+                img_path = extract_image_locations(error.attrib["message"])
+                if img_path:
+                    failed_answers.append((test_name, img_path))
+    return failed_answers, missing_answers
```
Comment (colinmarc) [inline tests/cloud_answer_tests_runner.py]: This whole thing is pretty unobjectionable, but I have two questions:

 - Is it possible to put actual boolean metadata into the test xml output via our plugin, so that we don't have to scan the error message?
 - Does nose provide any utilities for parsing its own output?

File: tests/cloud_answer_tests_runner.py [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+    # ANSI color codes
+    COLOR_BLUE = '\x1b[34;1m'
+    COLOR_CYAN = '\x1b[36;1m'
+    COLOR_RESET = '\x1b[0m'
+
+    ANSWER_STORE = ytcfg.get("yt", "test_storage_dir")
+    if not ANSWER_STORE or ANSWER_STORE == "/does/not/exist":
+        ANSWER_STORE = "answer-store"
+
+    NOSETEST_XML = "answer_testing_nosetests.xml"
```
Comment (colinmarc) [inline tests/cloud_answer_tests_runner.py]: Super minor, but it doesn't make that much sense to have a "constant" here and then pass it as the argument to `parse_nose_xml`. I'd just have the constant there.

File: tests/cloud_answer_tests_runner.py [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+
+    # first parameter is program name
+    # https://github.com/python/cpython/blob/master/Lib/unittest/main.py#L99
+    test_argv = [os.path.basename(__file__), '--with-answer-testing',
+                 '--nologcapture', '-s', '-d', '-v', '--local',
+                 '--attr=answer_test', '--local-dir=%s' % ANSWER_STORE,
+                 '--with-xunit', '--xunit-file=%s' % NOSETEST_XML, 'yt']
+    result = nose.run(argv=test_argv, addplugins=[AnswerTesting()], exit=False)
```
Comment (colinmarc) [inline tests/cloud_answer_tests_runner.py]: Can we remove the `--attr "!answer_test"` from the nose command above, and then have this script just run as an `after_failure` on travis and operate purely on the xml?

**Comment by git-abhishek** [comment]
Further, the travis build page looks like as shown below where 2 new jobs (`Python: 2.7 Answer Test`, `Python: 3.6 Answer Test`) are added in `Tests` stage and all the tests are named either `Unit Test` or `Answer Test`:

yt - travis ci

File: yt/fields/tests/test_fields_plugins.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (colinmarc) [reply yt/fields/tests/test_fields_plugins.py]: I'm ignoring this file then, but tell me if I shouldn't.

[the remark this answers is not in the document]
Comment (colinmarc) [reply yt/fields/tests/test_fields_plugins.py]: We sorted this out :) There are commits from un-merged PRs still in here though (but they'll rebase out cleanly).

File: .travis.yml [hunk 18]
```diff
@@ -65,35 +66,54 @@ jobs:
...
     - stage: tests
+      name: "Python: 3.6 Answer Test"
       python: 3.6
       env: FLAKE8=
-      script: coverage run $(which nosetests) -c nose.cfg yt
+      script: coverage run $(which nosetests) --attr "answer_test" --xunit-file ANS_NOSETEST_XML -c nose_answer.cfg
+      after_failure: python tests/report_failed_answers.py -f -m --xunit-file ANS_NOSETEST_XML
 
     - stage: tests
+      name: "Xcode: xcode7.3 Unit Test"
```
Comment (colinmarc) [inline .travis.yml]: This should maybe be `Unit Tests` (with an `s`) and `Answer Tests`.

**Comment by git-abhishek** [comment]
I have updated the PR by splitting answer tests execution (run now as command line) and its `after_failure` step to be the failure/missing answer upload run as a python script. 

- Tested with modified test case, so that answer test fails and uploading the failed answers ([job](https://travis-ci.org/yt-project/yt/jobs/411960414#L827))
- Tested with no answer in `answer-store` so that the job fails and uploads the new answers ([job](https://travis-ci.org/yt-project/yt/jobs/411960415#L868))

Finally, reverting the code (above changes) to pass all the tests.

File: .travis.yml (hunk 18 above)
Comment (git-abhishek) [reply .travis.yml]: fixed

**Comment by git-abhishek** [comment]
@yt-fido test this please.
